### DISCUSSION TITLE
How to let a user delete their own account?

### DISCUSSION THREAD

**Discussion by ProjectMoon** [supabase/supabase#1066] [Questions] [answered]
There does not seem to be a delete function/method on the user object. Is it possible to let a user delete their own account via stored procedure/function that queries auth.users? Is it smart to let users execute delete statements against their own record in auth.users using row level security?

**Comment by silentworks** [comment, marked as answer]
Yes you would have to do it with a stored procedure for now, but this is a good feature to request the supabase team to add. A similar question was asked by another poster https://github.com/supabase/supabase/discussions/250#discussioncomment-97164

**Reply by ProjectMoon** [reply]
Yeah, I did that. However the permission is denied by default. I guess I need to modify the row level security on the auth users table?

**Reply by silentworks** [reply]
I would advise against modifying the row level security on the `auth.users` table. Please post the code for your stored procedure here, please.

**Reply by ProjectMoon** [reply]
It's just a delete statement. Where id = auth.uid(). (Sorry, responding from phone)

Edit: it actually takes user ID as a parameter to the stored procedure.

**Reply by silentworks** [reply]
This is untested but try this:

```sql
create or replace function delete_user() 
returns void as
$$
  delete from auth.users 
  where id = auth.uid()
$$ 
language plpgsql;
```
Also, I wouldn't allow for passing the id as property unless it's for admin purpose.

**Reply by ProjectMoon** [reply]
The function I ended up creating is this:

```
CREATE or replace function delete_user()
  returns void
LANGUAGE SQL
AS $$
   delete from auth.users where id = auth.uid();
$$;
```

This unfortunately still gets permission denied errors.

**Reply by silentworks** [reply]
Ok I have updated my original function to include the `SECURITY DEFINER` and it works now.

```sql
CREATE or replace function delete_user()
  returns void
LANGUAGE SQL SECURITY DEFINER 
AS $$
   delete from auth.users where id = auth.uid();
$$;
```

Here is a short video showing this in action.

https://user-images.githubusercontent.com/79497/113702667-5b57b680-96c9-11eb-8011-9fb8286291dd.mp4

**Reply by ProjectMoon** [reply]
This works like a charm. Thanks.

Also had to make sure to sign out the session to completely shut everything down. Otherwise I ended up with a missing user row and an active user session.

**Reply by Rizialdi** [reply]
Awesome.
I was wondering about the `rpc syntax` and on the documentation I didn't see a clear explanation on how to write them.

I tried to pass a specific id:
```sql
CREATE or replace function "deleteUser"(id text)
  returns void
LANGUAGE SQL SECURITY DEFINER 
AS $$
   delete from users where users.id = id;
   delete from auth.users where id = id;
$$;
```
and calling it like that:
```js
await supabase
   .rpc('deleteUser', { id: user.id})
``` 

However many attempts, it still deletes all user records. 

Do you have an idea of what I'm doing wrongly ? Thanks in advance.

**Reply by inian** [reply]
` delete from auth.users where id = id;` is evaluating to ` delete from auth.users where true` I think. try giving the function parameter a different name or doing `delete from auth.users where auth.users.id = id;`?

**Reply by jasonho-lynx** [reply]
> This works like a charm. Thanks.
> 
> Also had to make sure to sign out the session to completely shut everything down. Otherwise I ended up with a missing user row and an active user session.

@ProjectMoon how were you able to sign out after deleting the user? I'm getting an error `{"message":"Invalid user","status":401}` when trying to use `await supabase.auth.signOut();` after calling the `deleteUser` `rpc`.

**Reply by ProjectMoon** [reply]
Just by calling the sign out method. Maybe I was not checking the error? Or perhaps the behavior has changed.

**Reply by brauliolledo** [reply]
This thread helped me a lot for my use case, so thanks!

To add to the solution: if I wanted to give the service_role the ability to update a user through the supabase API (for example, when building a backend service or a cloud function) and not having to use the pg client in such service and without using SECURITY DEFINER (which can be abused afaik), is to explicitly grant usage and granular (select, update, delete or insert) permissions to the service_role role on postgres' RLS. I'm not sure why, but the RLS bypass for the service_role set up on the initial schema creation (in the self-hosted version, at least) isn't working as expected.

Example
`grant usage on schema auth to service_role;
grant update, select on auth.users to service_role;`

So now, given you're authenticated with a service_role JWT, you can call the modifying function through an RPC.

I believe this also applies if we want to do this with the anon role: grant specific and granular permissions to the table we want to edit on the auth schema, and call the function through RPC.

**Reply by tonyborchert** [reply]
Not too sure about this, but wouldn't that be a total security flaw, as everyone could delete any account?

My alternative:
```sql
CREATE or replace function "deleteUser"()
  returns void
LANGUAGE SQL SECURITY DEFINER 
AS $$
   delete from auth.users where id = auth.uid();
$$;
```

Also its probably smart to design the `public.users` table like this:

```sql
create table public.users(
  id uuid references auth.users on delete cascade,
  ...
);
```

This would basically automatically delete all users of the public.users, if there isn't a user with the same uuid in auth.users

**Reply by SerkanToraman** [reply]
This function worked well for me, thanks

CREATE or replace function "deleteUser"()
  returns void
LANGUAGE SQL SECURITY DEFINER 
AS $$
   delete from auth.users where id = auth.uid();
$$;

**Comment by MadeByDouglas** [comment]
OK I am very new to Postgres and the like so I apologize if this is redundant, but I also have this question about deleting accounts. I'm building a Swift app and part of the App Store Policy is users must be able to delete their own account. So since I can't trigger this in the Swift Supabase API, is the right approach to use Database Functions or Edge Functions? 

And the suggestion above a "stored procedure" is that connected with RLS? Is that called a "Database Function" ? Is that still the correct approach or has the API changed since this discussion (almost 2 years ago)

I searched the Docs, even with the Clippy bot but didn't really get satisfactory answers

**Reply by arbyruns** [reply]
I'm in the same situation as you right now.

**Reply by mansueli** [reply]
Hey @arbyruns, nice to see you here! 🫡  We worked 'together' on Outlook Mobile Support. ( I was in Upwork support). 

You should be able to invoke an edge function like this from your Swift app:
https://blog.mansueli.com/supabase-user-self-deletion-empower-users-with-edge-functions

If you want them to double-check with a password, there's an example here:
https://blog.mansueli.com/secure-password-verification-and-update-with-supabase-and-postgresql

**Reply by arbyruns** [reply]
@mansueli oh wow, small world! Thanks for putting that blog together and hope you're doing well!

**Comment by MartinMalinda** [comment]
I think edge functions are a correct way to handle this. I think this is particularly tricky to cover securely just with RLS.

This is the flow I'm going for:

1) Create (authorized) edge handler for `/request-account-delete`. It will add a row to `account_delete_tokens` table and send a link to User's email.
2) The link points to something like `/confirmDelete?token=....`
3) Client reads the token from URL and sends a new request to `/confirm-account-delete?token=....` 
4) This (authorized) edge handler verifies the validity of the token and if all is well, deletes the user from the users table. Hopefully cascades take care of the rest.
5) If `200` client does `localStorage.clear()` and `window.reload()` and pops up some account delete success message

I think 1 - 4 could be provided by Supabase by default. After all Supabase is already sending tokens to email as part of magic link login and this is quite similar.

I think in ideal case it could be `await supabase.auth.requestUserDelete()` and the other part could be handled during supabase initialisation.

Deletion via `rpc` is problematic because that allows user to delete their account ASAP, without extra guards in check. Probably better not to have it so easy. Better not allow a potential XSS logic do stuff like this, or even someone physically using your computer shouldn't be able to delete your account with one simple click. Especially if it can't be undone. This is why these things should be "two-factor".

**Reply by mansueli** [reply]
I've written examples of this here:

https://blog.mansueli.com/supabase-user-self-deletion-empower-users-with-edge-functions

**Reply by canardos** [reply]
Although not a simple-one call API method, you can accomplish 1-4 using the `reauthenticate` method in the auth API.

This method is used for 2-factor authentication of user data changes etc, but can be repurposed for account deletion.

When you call `reauthenticate`, an OTP is generated and sent to the user's email. A SHA-224 hash is also generated from the concatenation of the user's email address and the OTP (as a string). This hash is stored in the `reauthentication_token` column of the `auth.users` table.

So you can:

1. Call the `reauthenticate` API method on the client side when the user chooses to delete their account.
2. Pop up a window asking for the OTP to confirm deletion.
3. Post the OTP to a `delete-user` edge function.
4. The edge function calculates the hash of the user's email address and the OTP and verifies it against the hash in the users table.
5. Delete the user if it matches, fail otherwise

**There are two complications.**

The first is that this relies on undocumented behavior of the API (specifically, where the hash is stored, the type of hash and how it is calculated). The API in general is quite poorly documented, so although it's not documented, it's not clear if this is intended to be used this way. **Hopefully someone from the team can weigh in.**

The second is getting access to the `reauthentication_token` column in the `auth.users` table. Unfortunately, the `auth.admin.getUserById` function does not return the reauth columns, and the `service_role` does not have select privileges on the table - only `supabase_auth_admin` and `dashboard_user` do.

The simplest (and probably safest) way around this is to put an update trigger on the `reauthentication_token` column, which copies the token to a table with RLS (i.e. user_profile if you have one) in the public schema.

Another option is to create a `SELECT id,reauthentication_token,reauthentication_sent_at FROM auth.users` view in the public schema, but you should restrict access to the `service_role` (i.e. `REVOKE ALL` from everyone else).

In general, I'd prefer the ability to reauthenticate by password so that the user could simply be required to reenter their password for account deletion, email change (since you might have lost the old address), user data change etc., but for some reason the API design only permits email.

**Reply by mansueli** [reply]
If you want to use their own passwords to confirm the deletion, then you can just combine the example above with this other one:

https://blog.mansueli.com/secure-password-verification-and-update-with-supabase-and-postgresql
